src/analyser.cpp

```diff
@@ -0,0 +1,1509 @@
+{
+    if (checkInitialValue && !variable->initialValue().empty()) {
+        // The variable has an initial value, so it can either be a constant or
+        // a state. By default, we consider it to be a constant and, if we find
+        // an ODE for that variable, we will know that it was actually a state.
+
+        mType = Type::CONSTANT;
+
+        mInitialisingVariable = variable;
+    }
+
+    mVariable = variable;
+}
+
+void AnalyserInternalVariable::makeVoi()
+{
+    mType = Type::VARIABLE_OF_INTEGRATION;
+}
+
+void AnalyserInternalVariable::makeState()
+{
+    if (mType == Type::UNKNOWN) {
+        mType = Type::SHOULD_BE_STATE;
+    } else if (mType == Type::CONSTANT) {
+        mType = Type::STATE;
+    }
+}
+
+#ifdef SWIG
+struct AnalyserEquation
+#else
+struct AnalyserInternalEquation: public std::enable_shared_from_this<AnalyserInternalEquation>
+#endif
+{
+    enum struct Type
+    {
+        UNKNOWN,
+        TRUE_CONSTANT,
+        VARIABLE_BASED_CONSTANT,
+        RATE,
+        ALGEBRAIC
+    };
+
+    size_t mOrder = MAX_SIZE_T;
+    Type mType = Type::UNKNOWN;
+
+    std::vector<AnalyserInternalEquationPtr> mDependencies;
+
+    AnalyserEquationAstPtr mAst;
+
+    std::vector<AnalyserInternalVariablePtr> mVariables;
+    std::vector<AnalyserInternalVariablePtr> mOdeVariables;
+
+    AnalyserInternalVariablePtr mVariable;
+    ComponentPtr mComponent = nullptr;
+
+    bool mComputedTrueConstant = true;
+    bool mComputedVariableBasedConstant = true;
+
+    bool mIsStateRateBased = false;
+
+    explicit AnalyserInternalEquation(const ComponentPtr &component);
+
+    void addVariable(const AnalyserInternalVariablePtr &variable);
+    void addOdeVariable(const AnalyserInternalVariablePtr &odeVariable);
+
+    static bool containsNonUnknownVariables(const std::vector<AnalyserInternalVariablePtr> &variables);
+    static bool containsNonConstantVariables(const std::vector<AnalyserInternalVariablePtr> &variables);
+
+    static bool knownVariable(const AnalyserInternalVariablePtr &variable);
+    static bool knownOdeVariable(const AnalyserInternalVariablePtr &odeVariable);
+
+    bool check(size_t & equationOrder, size_t & stateIndex, size_t & variableIndex);
+};
+
+AnalyserInternalEquation::AnalyserInternalEquation(const ComponentPtr &component)
+    : mAst(AnalyserEquationAst::create())
+    , mComponent(component)
+{
+}
+
+void AnalyserInternalEquation::addVariable(const AnalyserInternalVariablePtr &variable)
+{
+    if (std::find(mVariables.begin(), mVariables.end(), variable) == mVariables.end()) {
+        mVariables.push_back(variable);
+    }
+}
+
+void AnalyserInternalEquation::addOdeVariable(const AnalyserInternalVariablePtr &odeVariable)
+{
+    if (std::find(mOdeVariables.begin(), mOdeVariables.end(), odeVariable) == mOdeVariables.end()) {
+        mOdeVariables.push_back(odeVariable);
+    }
+}
+
+bool AnalyserInternalEquation::containsNonUnknownVariables(const std::vector<AnalyserInternalVariablePtr> &variables)
```

Comment:
This one makes me smile :) (see?). In other places we have used "has" in function names (in place of "contains") ... but I wondered about the double negative here ... maybe `hasKnownVariables` is clearer?

Reply:
Agreed about "has" vs. "contains" (same for `containsNonConstantVariables()` -> `hasNonConstantVariables()`). Also agreed about double negatives. I hate those and wonder how I allowed myself to use one! (Clearly just used the type of variable -- `UNKNOWN` -- in the function name!) Done in commit 41253532.